# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsVideoFrame.cpp
# Commit: 62d52e09851e
# Full Hash: 62d52e09851eeef665ca67ba97ee22139c174945
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2021-05-16 09:17:48
# Regressor Bug: 1540737
# File Overlap Count: 2
# Description:
#   Bug 1540737 - Implement Paint() for all display items, and nsDisplayList so that we can paint without using FrameLayerBuilder. r=miko
#   
#   This creates a new leaf nsDisplayWrapper item for consumers that just want an empty wrapper, with a Paint method that asserts. This lets us leaves Paint pure-virtual on nsDisplayWrapList, and lets the compiler force us to implement it for the other leaf types.
#   
#   It also disables flattening of transform/opacity items. This is because they flatten the items even when used with FlattenedDisplayItemIterator, even though this doesn't add markers to replace them (only FLB does this). We could in theory fix this, but I don't think FLB performance matters much any more.
# ==============================================================================

diff -r 00af531b9d94 -r 62d52e09851e layout/generic/nsVideoFrame.cpp
--- a/layout/generic/nsVideoFrame.cpp	Sat May 15 22:38:45 2021 +0000
+++ b/layout/generic/nsVideoFrame.cpp	Sat May 15 22:38:45 2021 +0000
@@ -16,6 +16,7 @@
 #include "mozilla/dom/HTMLVideoElement.h"
 #include "mozilla/dom/ShadowRoot.h"
 #include "mozilla/layers/RenderRootStateManager.h"
+#include "BasicLayers.h"
 #include "nsDisplayList.h"
 #include "nsGenericHTMLElement.h"
 #include "nsPresContext.h"
@@ -210,7 +211,9 @@
   container->SetScaleHint(scaleHint);
 
   RefPtr<ImageLayer> layer = static_cast<ImageLayer*>(
-      aManager->GetLayerBuilder()->GetLeafLayerFor(aBuilder, aItem));
+      aManager->GetLayerBuilder()
+          ? aManager->GetLayerBuilder()->GetLeafLayerFor(aBuilder, aItem)
+          : nullptr);
   if (!layer) {
     layer = aManager->CreateImageLayer();
     if (!layer) return nullptr;
@@ -497,6 +500,26 @@
     HTMLVideoElement* video = HTMLVideoElement::FromNode(f->GetContent());
     return video->VideoWidth() > 0;
   }
+
+  virtual void Paint(nsDisplayListBuilder* aBuilder,
+                     gfxContext* aCtx) override {
+    // This currently uses BasicLayerManager to re-use the code for extracting
+    // the current Image and generating DrawTarget rendering commands for it.
+    // Ideally we'll factor out that code and use it directly soon.
+    RefPtr<BasicLayerManager> layerManager =
+        new BasicLayerManager(BasicLayerManager::BLM_OFFSCREEN);
+
+    layerManager->BeginTransactionWithTarget(aCtx);
+    RefPtr<Layer> layer =
+        BuildLayer(aBuilder, layerManager, ContainerLayerParameters());
+    if (!layer) {
+      layerManager->AbortTransaction();
+      return;
+    }
+
+    layerManager->SetRoot(layer);
+    layerManager->EndEmptyTransaction();
+  }
 };
 
 void nsVideoFrame::BuildDisplayList(nsDisplayListBuilder* aBuilder,